zio-config/src/main/scala/zio/config/actions/Write.scala

```diff
@@ -13,8 +13,16 @@ object Write {
           Map(path -> propertyType.write(aa))
         })
 
-      case Config.Xmap(c, _, to) =>
-        Write(ZIO.accessM(b => write(c).run.provide(to(b))))
+      case Config.Optional(c) =>
+        Write(
+          ZIO.accessM(
+            a =>
+              a.asInstanceOf[Option[A]]
+                .fold[ZIO[A, WriteError, Map[String, String]]](
+                  ZIO.succeed(Map.empty[String, String])
+                )(aa => write(c).run.provide(aa))
```

Comment:
@jdegoes The place where I had to use `asInstanceOf`.

Reply:
didn't understand, the a is of type A right?
I would do this: (but not sure if it would be correct)
```scala
case c @ Config.Optional(_) =>
  write(c).run.fromOption.foldM(_ => UIO(Map.empty[String, String]), UIO(_))
```

Consider to add test

Reply:
The optional was already implemented before using `opt` function in package object.
Hence the tests were there before.

Reply:
I tried 
```
case c @ Config.Optional(_) =>
  write(c).run.fromOption.foldM(_ => UIO(Map.empty[String, String]), UIO(_))
```
It's actually interesting. But unfortunately the types are still not aligned. `A` comes in the contra variant position for write. And we say, the `A` was actually `Option[A]` during a `Config.Optional` program.

It's almost a natural transformation that we are doing, and with Scala, we are unaware of the inner types in GADT, and we have to explicitly do the cast.